File: secretspec-node/test/conformance.test.js
'use strict';

// Cross-language conformance: resolve the shared fixtures and assert this SDK
// produces the canonical result every other SDK must also produce.

const assert = require('node:assert');
const test = require('node:test');
const fs = require('node:fs');
const path = require('node:path');
const { execFileSync } = require('node:child_process');

function ensureLib() {
  if (process.env.SECRETSPEC_FFI_LIB) return;
  const repo = path.resolve(__dirname, '..', '..');
  execFileSync('cargo', ['build', '-p', 'secretspec-ffi'], { cwd: repo, stdio: 'inherit' });
  const meta = JSON.parse(
    execFileSync('cargo', ['metadata', '--no-deps', '--format-version', '1'], { cwd: repo }),
  );
  const name =
    process.platform === 'darwin' ? 'libsecretspec_ffi.dylib' : 'libsecretspec_ffi.so';
  process.env.SECRETSPEC_FFI_LIB = path.join(meta.target_directory, 'debug', name);
}

ensureLib();
const { SecretSpec } = require('../index.js');

const FIXTURES = path.resolve(__dirname, '..', '..', 'conformance', 'fixtures');

function canonical(resolved) {
  const secrets = {};
  for (const [name, secret] of Object.entries(resolved.secrets)) {
    const value = secret.asPath ? fs.readFileSync(secret.get(), 'utf8') : secret.value;
    secrets[name] = { value, source: secret.source, as_path: secret.asPath };
  }
  return {
    profile: resolved.profile,
    secrets,
    missing_required: [],
    missing_optional: [...resolved.missingOptional].sort(),
  };
}

for (const fixture of fs.readdirSync(FIXTURES).sort()) {
  const dir = path.join(FIXTURES, fixture);
  if (!fs.statSync(dir).isDirectory()) continue;

  test(`conformance: ${fixture}`, () => {
    const expected = JSON.parse(fs.readFileSync(path.join(dir, 'expected.json'), 'utf8'));
    const resolved = SecretSpec.builder()
      .withPath(path.join(dir, 'secretspec.toml'))
      .withProvider(`dotenv://${path.join(dir, '.env')}`)
      .withReason('conformance')
      .load();

    assert.deepStrictEqual(canonical(resolved), expected);
  });
}

File: secretspec-ex/test/conformance_test.exs
defmodule SecretSpec.ConformanceTest do
  use ExUnit.Case, async: true

  @fixtures Path.expand("../../conformance/fixtures", __DIR__)

  test "matches the canonical basic fixture" do
    fixture = Path.join(@fixtures, "basic")
    expected = fixture |> Path.join("expected.json") |> File.read!() |> Jason.decode!()
    manifest = Path.join(fixture, "secretspec.toml")
    provider = "dotenv://#{Path.join(fixture, ".env")}"

    resolved =
      SecretSpec.builder()
      |> SecretSpec.Builder.with_path(manifest)
      |> SecretSpec.Builder.with_provider(provider)
      |> SecretSpec.Builder.with_reason("conformance")
      |> SecretSpec.Builder.load()

    assert canonical(resolved) == expected
  end

  defp canonical(resolved) do
    secrets =
      Map.new(resolved.secrets, fn {name, secret} ->
        value = if secret.as_path, do: File.read!(secret.path), else: secret.value
        {name, %{"value" => value, "source" => secret.source, "as_path" => secret.as_path}}
      end)

    %{
      "profile" => resolved.profile,
      "secrets" => secrets,
      "missing_required" => [],
      "missing_optional" => Enum.sort(resolved.missing_optional)
    }
  end
end
